cpp/src/rolling/detail/rolling_utils.cuh

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/detail/iterator.cuh>
+#include <cudf/types.hpp>
+
+#include <cuda/functional>
+
+namespace cudf {
+
+namespace detail::rolling {
+
+/**
+ * @brief A group descriptor for an ungrouped rolling window.
+ *
+ * @param num_rows The number of rows to be rolled over.
+ *
+ * @note This is used for uniformity of interface between grouped and ungrouped iterator
+ * construction.
+ */
+struct ungrouped {
+  cudf::size_type num_rows;
+
+  [[nodiscard]] __device__ constexpr cudf::size_type label(cudf::size_type) const noexcept
+  {
+    return 0;
+  }
+  [[nodiscard]] __device__ constexpr cudf::size_type start(cudf::size_type) const noexcept
+  {
+    return 0;
+  }
+  [[nodiscard]] __device__ constexpr cudf::size_type end(cudf::size_type) const noexcept
+  {
+    return num_rows;
+  }
+};
+
+/**
+ * @brief A group descriptor for a grouped rolling window.
+ *
+ * @param labels The group labels, mapping from input rows to group.
+ * @param offsets The group offsets providing the endpoints of each group.
+ *
+ * @note This is used for uniformity of interface between grouped and ungrouped iterator
+ * construction.
+ */
+struct grouped {
+  cudf::size_type const* labels;
+  cudf::size_type const* offsets;
+
+  [[nodiscard]] __device__ constexpr cudf::size_type label(cudf::size_type i) const noexcept
+  {
+    return labels[i];
+  }
+  [[nodiscard]] __device__ constexpr cudf::size_type start(cudf::size_type label) const noexcept
+  {
+    return offsets[label];
+  }
+  [[nodiscard]] __device__ constexpr cudf::size_type end(cudf::size_type label) const noexcept
+  {
+    return offsets[label + 1];
+  }
+};
+
+enum class direction : bool {
+  PRECEDING,
+  FOLLOWING,
+};
+
+template <typename Grouping, direction Direction>
+struct fixed_window_clamper {
+  Grouping groups;
+  cudf::size_type delta;
+  [[nodiscard]] __device__ cudf::size_type operator()(cudf::size_type i)
```

Comment:
Do you suppose we might get away with making this a `const` function?
```suggestion
  [[nodiscard]] __device__ cudf::size_type operator()(cudf::size_type i) const
```

Reply:
Oh yes, thanks.

Reply:
Ah, it can't be noexcept because min/max are not noexcept